src/cdomain/value/cdomains/int/bitfieldDomain.ml

```diff
@@ -0,0 +1,623 @@
+    match nth_bit z n, nth_bit o n with
+    | true, true -> `Undetermined
+    | false, false -> `Invalid
+    | true, false -> `Zero
+    | false, true -> `One
+
+  let min ik (z,o) = 
+    let signBit = Ints_t.one <<: ((Size.bit ik) - 1) in 
+    let signMask = !: (Ints_t.of_bigint (snd (Size.range ik))) in
+    let isNegative = signBit &: o <> Ints_t.zero in
+    if isSigned ik && isNegative then Ints_t.to_bigint(signMask |: (!: z))
+    else Ints_t.to_bigint(!: z)
+
+  let max ik (z,o) =
+    let signBit = Ints_t.one <<: ((Size.bit ik) - 1) in 
+    let signMask = Ints_t.of_bigint (snd (Size.range ik)) in
+    let isPositive = signBit &: z <> Ints_t.zero in
+    if isSigned ik && isPositive then Ints_t.to_bigint(signMask &: o)
+    else Ints_t.to_bigint o 
+
+  let rec concretize (z,o) = (* O(2^n) *)
+    if is_const (z,o) then [o]
+    else
+      let bit = o &: Ints_t.one in
+      let bf_bit = nth_bf_bit (z,o) 0 in
+      concretize (z >>. 1, o >>: 1) |>
+      if bf_bit = `Undetermined then
+        List.concat_map (fun c -> [c <<: 1; (c <<: 1) |: Ints_t.one])
+      else if bf_bit = `Invalid then
+        failwith "Should not have happened: Invalid bit during concretization of a bitfield."
+      else
+        List.map (fun c -> c <<: 1 |: bit)
+
+  let concretize bf = List.map Ints_t.to_int (concretize bf)
+
+  let shift_right ik (z,o) c =
+    let msb_pos = (Size.bit ik - c) in
+    let msb_pos = if msb_pos < 0 then 0 else msb_pos in
+    let sign_mask = !:(bitmask_up_to msb_pos) in
+    if isSigned ik && o <: Ints_t.zero then
+      (z >>: c, (o >>: c) |: sign_mask)
+    else
+      ((z >>: c) |: sign_mask, o >>: c)
+
+  let shift_right ik bf (z2, o2) = 
+    if is_const (z2, o2) 
+    then 
+      shift_right ik bf (Ints_t.to_int o2)
+    else
+      let shift_counts = concretize (z2, o2) in
+      List.fold_left (fun acc c ->
+          let next = shift_right ik bf c in
+          join acc next
+        ) (zero_mask, zero_mask) shift_counts
+
+  let shift_left _ (z,o) c =
+    let zero_mask = bitmask_up_to c in
+    ((z <<: c) |: zero_mask, o <<: c)
+
+  let shift_left ik bf (z2, o2) = 
+    if is_const (z2, o2) 
+    then 
+      shift_left ik bf (Ints_t.to_int o2)
+    else
+      let shift_counts = concretize (z2, o2) in
+      List.fold_left (fun acc c ->
+          let next = shift_left ik bf c in
+          join acc next
+        ) (zero_mask, zero_mask) shift_counts
+
+  let nth_bit p n = if nth_bit p n then Ints_t.one else Ints_t.zero
+end
+
+module BitfieldFunctor (Ints_t : IntOps.IntOps): SOverflow with type int_t = Ints_t.t and type t = (Ints_t.t * Ints_t.t) = struct
+
+  include InfixIntOps (Ints_t)
+
+  let name () = "bitfield"
+  type int_t = Ints_t.t
+  type t = (Ints_t.t * Ints_t.t) [@@deriving eq, ord, hash]
+
+  module BArith = BitfieldArith (Ints_t)
+
+  let top () = (BArith.one_mask, BArith.one_mask)
+  let bot () = (BArith.zero_mask, BArith.zero_mask)
+
+  let top_of ik = 
+    if isSigned ik then top () 
```

Comment:
I don't think i understand how you handle `ikinds` in general here.

- Do you norm the upper bits?
- If not do you not risk having values that are semantically equal but syntactically not? 
- What happens for `ikind`s that are wider than the `BArith.one_mask`?

Why can this not work for signed types? And does this not mean you find out nothing as soon as any computations happen in the program (even for unsigned types)?

C99 6.3.1.

> If an int can represent all values of the original type, the value is converted to an int; otherwise, it is converted to an unsigned int. These are called the integer
promotions. All other types are unchanged by the integer promotions.